File: android/app/src/main/kotlin/app/padscreen/protocol/WireProtocol.kt
package app.padscreen.protocol

import java.nio.ByteBuffer
import java.nio.ByteOrder

enum class MessageType(val code: Int) {
    HELLO(1),
    SESSION(2),
    CODEC_CONFIGURATION(3),
    VIDEO(4),
    INPUT(5),
    HEARTBEAT(6),
    ERROR(7);

    companion object {
        fun fromCode(code: Int): MessageType? = entries.firstOrNull { it.code == code }
    }
}

data class WireFrame(val type: MessageType, val payload: ByteArray) {
    override fun equals(other: Any?): Boolean =
        other is WireFrame && type == other.type && payload.contentEquals(other.payload)

    override fun hashCode(): Int = 31 * type.hashCode() + payload.contentHashCode()
}

sealed class ProtocolException(message: String) : Exception(message) {
    class InvalidMagic : ProtocolException("Invalid PDS1 magic")
    class UnsupportedVersion(val version: Int) : ProtocolException("Unsupported protocol version: $version")
    class UnknownMessageType(val code: Int) : ProtocolException("Unknown message type: $code")
    class PayloadTooLarge(val size: Int) : ProtocolException("Payload exceeds configured maximum: $size")
}

object FrameCodec {
    const val VERSION = 1
    const val HEADER_SIZE = 12
    private val magic = byteArrayOf(0x50, 0x44, 0x53, 0x31)

    fun encode(frame: WireFrame): ByteArray {
        val output = ByteBuffer.allocate(HEADER_SIZE + frame.payload.size).order(ByteOrder.BIG_ENDIAN)
        output.put(magic)
        output.put(VERSION.toByte())
        output.put(frame.type.code.toByte())
        output.putShort(0)
        output.putInt(frame.payload.size)
        output.put(frame.payload)
        return output.array()
    }
}

class FrameDecoder(private val maxPayloadSize: Int = 16 * 1024 * 1024) {
    private var buffer = ByteArray(64 * 1024)
    private var readOffset = 0
    private var writeOffset = 0

    fun append(bytes: ByteArray): List<WireFrame> = append(bytes, bytes.size)

    fun append(bytes: ByteArray, length: Int): List<WireFrame> {
        require(length in 0..bytes.size)
        ensureWritable(length)
        bytes.copyInto(buffer, writeOffset, 0, length)
        writeOffset += length
        val frames = mutableListOf<WireFrame>()

        while (writeOffset - readOffset >= FrameCodec.HEADER_SIZE) {
            if (buffer[readOffset] != 0x50.toByte() ||
                buffer[readOffset + 1] != 0x44.toByte() ||
                buffer[readOffset + 2] != 0x53.toByte() ||
                buffer[readOffset + 3] != 0x31.toByte()
            ) {
                throw ProtocolException.InvalidMagic()
            }
            val version = buffer[readOffset + 4].toInt() and 0xff
            if (version != FrameCodec.VERSION) throw ProtocolException.UnsupportedVersion(version)
            val rawType = buffer[readOffset + 5].toInt() and 0xff
            val type = MessageType.fromCode(rawType) ?: throw ProtocolException.UnknownMessageType(rawType)
            val payloadSize = ((buffer[readOffset + 8].toInt() and 0xff) shl 24) or
                ((buffer[readOffset + 9].toInt() and 0xff) shl 16) or
                ((buffer[readOffset + 10].toInt() and 0xff) shl 8) or
                (buffer[readOffset + 11].toInt() and 0xff)
            if (payloadSize < 0 || payloadSize > maxPayloadSize) {
                throw ProtocolException.PayloadTooLarge(payloadSize)
            }
            val frameSize = FrameCodec.HEADER_SIZE + payloadSize
            if (writeOffset - readOffset < frameSize) break

            val payloadStart = readOffset + FrameCodec.HEADER_SIZE
            frames += WireFrame(type, buffer.copyOfRange(payloadStart, payloadStart + payloadSize))
            readOffset += frameSize
        }

        compactIfUseful()
        return frames
    }

    private fun ensureWritable(additionalBytes: Int) {
        if (buffer.size - writeOffset >= additionalBytes) return
        val unreadBytes = writeOffset - readOffset
        if (buffer.size - unreadBytes >= additionalBytes) {
            buffer.copyInto(buffer, 0, readOffset, writeOffset)
            readOffset = 0
            writeOffset = unreadBytes
            return
        }

        var capacity = buffer.size
        val required = unreadBytes + additionalBytes
        while (capacity < required) capacity = (capacity * 2).coerceAtLeast(required)
        val replacement = ByteArray(capacity)
        buffer.copyInto(replacement, 0, readOffset, writeOffset)
        buffer = replacement
        readOffset = 0
        writeOffset = unreadBytes
    }

    private fun compactIfUseful() {
        if (readOffset == writeOffset) {
            readOffset = 0
            writeOffset = 0
        } else if (readOffset >= buffer.size / 2) {
            val unreadBytes = writeOffset - readOffset
            buffer.copyInto(buffer, 0, readOffset, writeOffset)
            readOffset = 0
            writeOffset = unreadBytes
        }
    }
}

File: Sources/PadScreenProtocol/WireProtocol.swift
import Foundation

public enum MessageType: UInt8, Codable, Sendable {
    case hello = 1
    case session = 2
    case codecConfiguration = 3
    case video = 4
    case input = 5
    case heartbeat = 6
    case error = 7
}

public struct WireFrame: Equatable, Sendable {
    public let type: MessageType
    public let payload: Data

    public init(type: MessageType, payload: Data) {
        self.type = type
        self.payload = payload
    }
}

public enum ProtocolError: Error, Equatable, Sendable {
    case invalidMagic
    case unsupportedVersion(UInt8)
    case unknownMessageType(UInt8)
    case payloadTooLarge(Int)
}

public enum FrameCodec {
    public static let version: UInt8 = 1
    public static let headerSize = 12
    private static let magic: [UInt8] = [0x50, 0x44, 0x53, 0x31]

    public static func encode(_ frame: WireFrame) throws -> Data {
        guard frame.payload.count <= Int(UInt32.max) else {
            throw ProtocolError.payloadTooLarge(frame.payload.count)
        }

        var output = Data(magic)
        output.append(version)
        output.append(frame.type.rawValue)
        output.append(contentsOf: [0, 0])
        let length = UInt32(frame.payload.count)
        output.append(contentsOf: [
            UInt8((length >> 24) & 0xff),
            UInt8((length >> 16) & 0xff),
            UInt8((length >> 8) & 0xff),
            UInt8(length & 0xff),
        ])
        output.append(frame.payload)
        return output
    }
}

public struct FrameDecoder: Sendable {
    private var buffer = Data()
    private let maxPayloadSize: Int

    public init(maxPayloadSize: Int = 16 * 1024 * 1024) {
        self.maxPayloadSize = maxPayloadSize
    }

    public mutating func append<S: Sequence>(_ bytes: S) throws -> [WireFrame] where S.Element == UInt8 {
        buffer.append(contentsOf: bytes)
        var frames: [WireFrame] = []

        while buffer.count >= FrameCodec.headerSize {
            guard Array(buffer.prefix(4)) == [0x50, 0x44, 0x53, 0x31] else {
                throw ProtocolError.invalidMagic
            }
            let version = buffer[4]
            guard version == FrameCodec.version else {
                throw ProtocolError.unsupportedVersion(version)
            }
            let rawType = buffer[5]
            guard let type = MessageType(rawValue: rawType) else {
                throw ProtocolError.unknownMessageType(rawType)
            }
            let payloadLength = (Int(buffer[8]) << 24)
                | (Int(buffer[9]) << 16)
                | (Int(buffer[10]) << 8)
                | Int(buffer[11])
            guard payloadLength <= maxPayloadSize else {
                throw ProtocolError.payloadTooLarge(payloadLength)
            }
            let frameLength = FrameCodec.headerSize + payloadLength
            guard buffer.count >= frameLength else { break }

            let payload = Data(buffer[FrameCodec.headerSize..<frameLength])
            frames.append(WireFrame(type: type, payload: payload))
            buffer = Data(buffer.dropFirst(frameLength))
        }

        return frames
    }
}

public struct HelloMessage: Codable, Equatable, Sendable {
    public let viewportWidth: Int
    public let viewportHeight: Int
    public let density: Double
    public let codecs: [String]

    public init(viewportWidth: Int, viewportHeight: Int, density: Double, codecs: [String]) {
        self.viewportWidth = viewportWidth
        self.viewportHeight = viewportHeight
        self.density = density
        self.codecs = codecs
    }
}

public struct SessionMessage: Codable, Equatable, Sendable {
    public let sessionID: UUID
    public let width: Int
    public let height: Int
    public let framesPerSecond: Int
    public let codec: String

    public init(sessionID: UUID, width: Int, height: Int, framesPerSecond: Int, codec: String) {
        self.sessionID = sessionID
        self.width = width
        self.height = height
        self.framesPerSecond = framesPerSecond
        self.codec = codec
    }
}

public enum PointerAction: String, Codable, Equatable, Sendable {
    case down
    case move
    case up
    case cancel
    case scroll
}

public struct PointerMessage: Codable, Equatable, Sendable {
    public let action: PointerAction
    public let normalizedX: Double
    public let normalizedY: Double
    public let deltaX: Double
    public let deltaY: Double

    public init(action: PointerAction, normalizedX: Double, normalizedY: Double, deltaX: Double, deltaY: Double) {
        self.action = action
        self.normalizedX = normalizedX
        self.normalizedY = normalizedY
        self.deltaX = deltaX
        self.deltaY = deltaY
    }
}

public extension JSONEncoder {
    static var padScreen: JSONEncoder {
        let encoder = JSONEncoder()
        encoder.outputFormatting = [.sortedKeys]
        return encoder
    }
}

public extension JSONDecoder {
    static var padScreen: JSONDecoder { JSONDecoder() }
}
